### ISSUE TITLE
Fix worker.py concurrency issues.

### ISSUE THREAD

**Issue by robertnishihara** [ray-project/ray#3229]
We should use the `thread_safe_client` abstraction to protect the global worker and make all Ray API calls thin wrappers around calls to the global worker's methods.

For example, `ray.put` is currently not thread safe because `put_index` is not protected.

cc @stephanie-wang @suquark @guoyuhong

**Comment by guoyuhong** [comment]
I agree. We met a scenario that the some users hold a Java raylet client but submit tasks in multi-thread code. Then, the frontier of the actor is a mess and the dead lock of dummy objects will stop the job.
